platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/service/WritableRosterStore.java

```diff
@@ -0,0 +1,203 @@
+import com.swirlds.platform.state.RosterStateModifier;
+import com.swirlds.state.spi.CommittableWritableStates;
+import com.swirlds.state.spi.WritableKVState;
+import com.swirlds.state.spi.WritableSingletonState;
+import com.swirlds.state.spi.WritableStates;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import edu.umd.cs.findbugs.annotations.Nullable;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Objects;
+
+/**
+ * Read-write interface for accessing rosters states.
+ */
+public class WritableRosterStore implements RosterStateModifier {
+
+    private final WritableStates writableStates;
+    private final RosterStateAccessor rosterStateAccessor;
+
+    /**
+     * The roster state singleton. This is the state that holds the candidate roster hash and the list of pairs of
+     * active roster hashes and the round number in which those rosters became active.
+     *
+     * @implNote the use of {@link ReadablePlatformStateStore} and {@link WritablePlatformStateStore} to provide access
+     * to the roster states (beyond just the {@link PlatformState}) is deliberate, for convenience.
+     */
+    private final WritableSingletonState<RosterState> rosterState;
+
+    private final WritableKVState<ProtoBytes, Roster> rosterMap;
+
+    /**
+     * The maximum number of active rosters to keep in the roster state.
+     */
+    static final int MAXIMUM_ROSTER_HISTORY_SIZE = 2;
+
+    /**
+     * Constructs a new {@link WritableRosterStore} instance.
+     *
+     * @param writableStates the readable states
+     */
+    public WritableRosterStore(@NonNull final WritableStates writableStates) {
+        this.writableStates = writableStates;
+        this.rosterStateAccessor = new ReadableRosterStore(writableStates);
+        this.rosterState = writableStates.getSingleton(RosterStateId.ROSTER_STATES_KEY);
+        this.rosterMap = writableStates.get(RosterStateId.ROSTER_KEY);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void setCandidateRoster(@NonNull final Roster candidateRoster) {
+        Objects.requireNonNull(candidateRoster);
+        RosterValidator.validate(candidateRoster);
+
+        final Bytes incomingCandidateRosterHash =
+                RosterUtils.hash(candidateRoster).getBytes();
+
+        // update the roster state/map
+        final RosterState previousRosterState = rosterStateOrThrow();
+        final Bytes previousCandidateRosterHash = previousRosterState.candidateRosterHash();
+        final Builder newRosterState =
+                previousRosterState.copyBuilder().candidateRosterHash(incomingCandidateRosterHash);
+        removeRoster(previousCandidateRosterHash);
+        storeRoster(candidateRoster, incomingCandidateRosterHash, newRosterState);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Nullable
+    @Override
+    public Roster getCandidateRoster() {
+        return rosterStateAccessor.getCandidateRoster();
```

Comment:
This will not work. The `rosterStateAccessor` will not be aware of any changes until they are committed. Changes have to be kept in memory until they are committed by the workflow (e.g. at the very end of transaction handling).

Reply:
I think the API is supposed to return what it finds committed in the state. If a newly set roster is not yet committed, this API caller should handle this according to their business logic. I've just updated the javadoc to state the fact the roster might be unavailable immediately after been set.

Reply:
This is extremely unexpected behavior. Getters and setters have a long tradition in Java. Updating the javadocs is not sufficient IMO. If you really want the old value, you should rename this method.

Reply:
Are `getCandidateRosterIfCommitted()` and `setCandidateRosterForCommit()` as method names, with their respective javadocs better?

Reply:
In a similar scenario, we used `getOriginalCandidateRoster()`. I do not have a strong opinion about the name other than not calling them like regular getters and setters.

Reply:
The code has since been updated and this appears to be resolved.